File: spicedb-java/lib/src/main/java/com/authzed/spicedb/PermissionTree.java
package com.authzed.spicedb;

import java.util.List;

/**
 * A native node of an expanded permission tree.
 *
 * <p>Avoids leaking the proto {@code PermissionRelationshipTree} type in favor of plain Java
 * records. Returned by {@link SpiceDBClient#expandPermissionTree}. Exactly one of {@link
 * #intermediate} or {@link #leaf} is non-null.
 *
 * @param expandedObject the resource or subject object this node describes
 * @param expandedRelation the relation being expanded
 * @param intermediate non-null when this node combines child subtrees with a set operation
 * @param leaf non-null when this node holds the concrete subjects with access
 */
public record PermissionTree(
    ObjectRef expandedObject,
    String expandedRelation,
    IntermediateNode intermediate,
    LeafNode leaf) {

  /** The set operation combining an {@link IntermediateNode}'s children. */
  public enum Operation {
    UNSPECIFIED,
    UNION,
    INTERSECTION,
    EXCLUSION
  }

  /** Identifies a resource or subject object. */
  public record ObjectRef(String objectType, String objectID) {}

  /** A subject with access at a leaf of a {@link PermissionTree}. */
  public record SubjectRef(String subjectType, String subjectID, String optionalRelation) {}

  /** Combines child subtrees with a set operation. */
  public record IntermediateNode(Operation operation, List<PermissionTree> children) {}

  /** Holds the concrete subjects at a leaf of a {@link PermissionTree}. */
  public record LeafNode(List<SubjectRef> subjects) {}
}

File: spicedb-csharp/SpiceDB.Client/PermissionTree.cs
// PermissionTree is a native, immutable representation of an expanded
// permission tree. It avoids leaking the proto PermissionRelationshipTree
// type in favor of plain C# records.

namespace SpiceDB.Client;

/// <summary>
/// The set operation combining an <see cref="IntermediateNode"/>'s children.
/// </summary>
public enum TreeOperation
{
    Unspecified,
    Union,
    Intersection,
    Exclusion,
}

/// <summary>Identifies a resource or subject object.</summary>
public sealed record ObjectRef
{
    public string ObjectType { get; init; } = "";
    public string ObjectID { get; init; } = "";
}

/// <summary>A subject with access at a leaf of a <see cref="PermissionTree"/>.</summary>
public sealed record SubjectRef
{
    public string SubjectType { get; init; } = "";
    public string SubjectID { get; init; } = "";
    public string OptionalRelation { get; init; } = "";
}

/// <summary>Combines child subtrees with a set operation.</summary>
public sealed record IntermediateNode
{
    public TreeOperation Operation { get; init; }
    public IReadOnlyList<PermissionTree> Children { get; init; } = [];
}

/// <summary>Holds the concrete subjects at a leaf of a <see cref="PermissionTree"/>.</summary>
public sealed record LeafNode
{
    public IReadOnlyList<SubjectRef> Subjects { get; init; } = [];
}

/// <summary>
/// A native node of an expanded permission tree. Returned by
/// <see cref="SpiceDBClient.ExpandPermissionTreeAsync"/>. Exactly one of
/// <see cref="Intermediate"/> or <see cref="Leaf"/> is non-null.
/// </summary>
public sealed record PermissionTree
{
    public ObjectRef ExpandedObject { get; init; } = new();
    public string ExpandedRelation { get; init; } = "";
    public IntermediateNode? Intermediate { get; init; }
    public LeafNode? Leaf { get; init; }
}
